### PR TITLE
Feature/240 generic date validation

### PR DESCRIPTION
**Pull request by d-a-v-e**
## Changes in this PR

- [x] tidies up procurement detail controller
- [x] contract form object always expects an int for duration
- [x] generic date validation for all dates (starting with procurement start/end date attributes)

### CODE REVIEW

**Comment by d-a-v-e** [comment]
[Generic date validation](https://trello.com/c/U9OI4P7o/240-generic-date-validation)

File: app/forms/support/form.rb [hunk 1]
```diff
@@ -9,6 +9,23 @@ class Form
     #   @return [Hash] field validation error messages
     option :messages, Types::Hash, default: proc { {} }
 
+    # Convert a date hash in the format { year: #, month: #, day: # } into a date
```
Comment (peterdavidhamilton) [inline app/forms/support/form.rb:12]: Can I suggest that we adopt the convention of building custom types in the `::Types` namespace now. It'll help keep things organised and reusable (especially across the apps).

**Comment by d-a-v-e** [comment]
@PiperS52 @EdwinKruglov I'm unable to approve as I opened the PR but this all looks spot on 👍

File: app/controllers/support/concerns/has_date_params.rb [hunk 2]
```diff
@@ -13,11 +13,14 @@ module HasDateParams
...
-      # @return [Date]
+      # @return [Hash]
       def date_param(form_param, date_field)
         date = params.require(form_param).permit(date_field)
-        date_hash = { day: date["#{date_field}(3i)"], month: date["#{date_field}(2i)"], year: date["#{date_field}(1i)"] }
-        format_date(date_hash)
+        { day: date["#{date_field}(3i)"], month: date["#{date_field}(2i)"], year: date["#{date_field}(1i)"] }
+      end
+
+      def date_params(name)
```
Comment (peterdavidhamilton) [inline app/controllers/support/concerns/has_date_params.rb:22]: Quick comment

**Review by peterdavidhamilton — APPROVED** [review]
Check out the other branch before you merge if you can pick up where I've left off

**Pull request merged**